File: client/src/app/formulas/testing/test-cells.ts
import { tryParseCellAddress } from '../cell-ref';
import { CellReader } from '../cell-reader';
import { EMPTY, FormulaValue, valueFromRaw } from '../formula-value';

/** Test-only. A sheet made of a Map: anything not in it is empty. Counts reads for laziness tests. */
export class TestCells implements CellReader {
  private readonly cells = new Map<string, FormulaValue>();
  reads = 0;

  constructor(readonly dimensions: { readonly rows: number; readonly cols: number } = { rows: 100_000, cols: 26 }) {}

  /** Puts a constant in a cell using A1 notation, read the way a typed cell is read. */
  raw(address: string, raw: string | null): this {
    return this.value(address, valueFromRaw(raw));
  }

  value(address: string, value: FormulaValue): this {
    const position = tryParseCellAddress(address);
    if (!position) throw new Error(`bad address ${address}`);
    this.cells.set(`${position.row},${position.col}`, value);
    return this;
  }

  getValue(row: number, col: number): FormulaValue {
    this.reads++;
    return this.cells.get(`${row},${col}`) ?? EMPTY;
  }
}

File: server/tests/GridSync.Core.Tests/Formulas/TestCells.cs
using GridSync.Core.Formulas;

namespace GridSync.Core.Tests.Formulas;

/// <summary>A sheet made of a dictionary: anything not in it is empty. Counts reads for laziness tests.</summary>
internal sealed class TestCells(SheetDimensions dimensions) : ICellReader
{
    private readonly Dictionary<(int Row, int Col), FormulaValue> _cells = [];

    public SheetDimensions Dimensions { get; } = dimensions;

    public int Reads { get; private set; }

    public TestCells() : this(new SheetDimensions(100_000, 26)) { }

    /// <summary>Puts a constant in a cell using A1 notation, read the way a typed cell is read.</summary>
    public TestCells Raw(string address, string? raw) => Value(address, FormulaValue.FromRaw(raw));

    public TestCells Value(string address, FormulaValue value)
    {
        Assert.True(CellAddress.TryParse(address, out var row, out var col), $"bad address {address}");
        _cells[(row, col)] = value;
        return this;
    }

    public FormulaValue GetValue(int row, int col)
    {
        Reads++;
        return _cells.TryGetValue((row, col), out var value) ? value : FormulaValue.Empty;
    }
}
